custom/nic_compliance/utils.py

```diff
@@ -0,0 +1,119 @@
+                    return ''
+            else:
+                # this sounds like someone tried to obfuscate password but failed so ignore the password submitted
+                # completely
+                return ''
+        else:
+            # this sounds like someone tried to obfuscate password but failed so ignore the password submitted
+            # completely
+            return ''
+    else:
+        # return the password received AS-IS
+        return obfuscated_password
+
+
+def hash_password(password):
+    return PASSWORD_HASHER.encode(password, PASSWORD_HASHER.salt())
+
+
+def verify_password(password, password_salt):
+    return PASSWORD_HASHER.verify(password, password_salt)
+
+
+def obfuscated_passwords_redis_key_for_user(username):
+    return REDIS_LOGIN_ATTEMPTS_LIST_PREFIX + username
+
+
+def get_obfuscated_passwords(username):
+    client = get_redis_client()
+    return client.get(obfuscated_passwords_redis_key_for_user(username), [])
+
+
+def get_raw_password(obfuscated_password, username=None):
+    def replay_attack():
+        # Replay attack where the same obfuscated password used from previous login attempt
+        obfuscated_passwords = get_obfuscated_passwords(username)
+        for submitted_obfuscated_password in obfuscated_passwords:
+            if verify_password(obfuscated_password, submitted_obfuscated_password):
+                return True
+
+    def record_login_attempt():
+        client = get_redis_client()
+        key_name = obfuscated_passwords_redis_key_for_user(username)
+        obfuscated_passwords = client.get(key_name, [])
+        client.set(key_name, obfuscated_passwords + [hash_password(obfuscated_password)])
+        client.expire(key_name, timedelta(EXPIRE_LOGIN_ATTEMPTS_IN))
+
+    def _decode_password():
+        # force check for replay attack and recording login attempt only for web sign in by checking for username
+        # Also skip those two checks in case of 2-step authentication by checking for auth-username which is not
+        # present in consecutive token step's POST params
+        if username and request and request.POST.get('auth-username'):
+            if replay_attack():
+                return ''
+            record_login_attempt()
+        return extract_password(obfuscated_password)
+
+    if settings.OBFUSCATE_PASSWORD_FOR_NIC_COMPLIANCE:
+        request = get_request()
+        if request:
+            # 1. an attempt to decode a password should be done just once in a request for the login attempt
+            # check to work correctly and not consider it a replay attack in case of multiple calls
+            # 2. also there should be no need to decode a password multiple times in the same request.
+            if not hasattr(request, 'decoded_password'):
+                request.decoded_password = {}
+
+            # return decoded password set on request object for the obfuscated_password
+            if request.decoded_password.get(obfuscated_password):
```

Comment:
Since `_decode_password()` can return an empty string this should be `if obfuscated_password in request.decoded_password`